File: python/백준/다이나믹 프로그래밍/30805 사전 순 최대 공통 부분 수열.py
import sys
input = sys.stdin.readline

N = int(input())
A = list(map(int, input().split()))
M = int(input())
B = list(map(int, input().split()))

# 현재 A, B에서 어디까지 썼는지 가리키는 포인터
ia = 0
ib = 0
ans = []

# 숫자 범위가 1~100
MAX_VAL = 100

while ia < N and ib < M:
    chosen = None
    next_ia = next_ib = None

    # 이번에 뽑을 값 v를 큰 것부터 내려가며 탐색
    for v in range(MAX_VAL, 0, -1):
        # A에서 v가 ia 이후에 처음 나오는 위치 찾기
        pa = -1
        for i in range(ia, N):
            if A[i] == v:
                pa = i
                break
        if pa == -1:
            continue  # A에는 더 이상 v가 없음

        # B에서도 v가 ib 이후에 처음 나오는 위치 찾기
        pb = -1
        for j in range(ib, M):
            if B[j] == v:
                pb = j
                break
        if pb == -1:
            continue  # B에는 더 이상 v가 없음

        # A, B 둘 다에서 ia/ib 이후에 v를 찾았으니, 이 v를 선택
        chosen = v
        next_ia = pa + 1
        next_ib = pb + 1
        break  # 가장 큰 v를 찾았으니 더 내려갈 필요 없음

    if chosen is None:  # 더 이상 공통으로 뽑을 수 있는 수가 없음
        break

    ans.append(chosen)
    ia = next_ia
    ib = next_ib

print(len(ans))
if ans:
    print(*ans)

File: cpp/백준/다이나믹 프로그래밍/30805 사전 순 최대 공통 부분 수열.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    cin >> N;
    vector<int> A(N);
    for (int i = 0; i < N; ++i) cin >> A[i];

    int M;
    cin >> M;
    vector<int> B(M);
    for (int i = 0; i < M; ++i) cin >> B[i];

    int ia = 0, ib = 0;
    vector<int> ans;

    const int MAX_VAL = 100;

    while (ia < N && ib < M) {
        int chosen = -1;
        int next_ia = -1, next_ib = -1;

        for (int v = MAX_VAL; v >= 1; --v) {
            int pa = -1;
            for (int i = ia; i < N; ++i) {
                if (A[i] == v) {
                    pa = i;
                    break;
                }
            }
            if (pa == -1) continue;

            int pb = -1;
            for (int j = ib; j < M; ++j) {
                if (B[j] == v) {
                    pb = j;
                    break;
                }
            }
            if (pb == -1) continue;

            chosen = v;
            next_ia = pa + 1;
            next_ib = pb + 1;
            break;
        }

        if (chosen == -1) break;

        ans.push_back(chosen);
        ia = next_ia;
        ib = next_ib;
    }

    cout << ans.size() << "\n";
    if (!ans.empty()) {
        for (size_t i = 0; i < ans.size(); ++i) {
            cout << ans[i];
            if (i + 1 < ans.size()) cout << ' ';
        }
        cout << "\n";
    }

    return 0;
}
